dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/TrackedEntityAttributeControllerTest.java

```diff
@@ -0,0 +1,231 @@
+        programInstance.setFollowup( true );
+        manager.save( programInstance );
+    }
+
+    @Test
+    void getIndexableOnlyAttributes()
+    {
+
+        JsonObject json = GET( "/trackedEntityAttributes?indexableOnly=true" ).content( HttpStatus.OK );
+
+        assertFalse( json.isEmpty() );
+
+        Set<String> expectedAttributeSet = new HashSet<>();
+        expectedAttributeSet.add( "\"AttributeA\"" );
+        expectedAttributeSet.add( "\"AttributeB\"" );
+
+        assertAttributeList( json, expectedAttributeSet );
+
+    }
+
+    @Test
+    void getAllAttributes()
+    {
+
+        JsonObject json = GET( "/trackedEntityAttributes?indexableOnly=false" ).content( HttpStatus.OK );
+
+        assertFalse( json.isEmpty() );
+
+        Set<String> expectedAttributeSet = new HashSet<>();
+        expectedAttributeSet.add( "\"AttributeA\"" );
+        expectedAttributeSet.add( "\"AttributeB\"" );
+        expectedAttributeSet.add( "\"AttributeC\"" );
+        expectedAttributeSet.add( "\"AttributeD\"" );
+        expectedAttributeSet.add( "\"AttributeE\"" );
+
+        assertAttributeList( json, expectedAttributeSet );
+    }
+
+    @Test
+    void getIndexableAttributesAndFilterByIdShouldThrowError()
+    {
+        assertEquals( "indexableOnly parameter cannot be set if a separate filter for id is specified",
+            GET( "/trackedEntityAttributes?indexableOnly=true&filter=id:eq:ImspTQPwCqd" ).error(
+                HttpStatus.BAD_REQUEST ).getMessage() );
+    }
+
+    @Test
+    void getIndexableAttributesAndFilterByOtherParameters()
+    {
+        JsonObject json = GET(
+            "/trackedEntityAttributes?indexableOnly=true&filter=name:in:[AttributeB,AttributeC]" ).content(
+                HttpStatus.OK );
+
+        assertFalse( json.isEmpty() );
+
+        Set<String> expectedAttributeSet = new HashSet<>();
+        expectedAttributeSet.add( "\"AttributeB\"" );
+
+        assertAttributeList( json, expectedAttributeSet );
+    }
+
+    @Test
+    void getAttributesWithNameFilter()
+    {
+        JsonObject json = GET(
+            "/trackedEntityAttributes?indexableOnly=false&filter=name:in:[AttributeB,AttributeC]" ).content(
+                HttpStatus.OK );
+
+        assertFalse( json.isEmpty() );
+
+        Set<String> expectedAttributeSet = new HashSet<>();
+        expectedAttributeSet.add( "\"AttributeB\"" );
+        expectedAttributeSet.add( "\"AttributeC\"" );
+
+        assertAttributeList( json, expectedAttributeSet );
+    }
+
+    private static void assertAttributeList( JsonObject actualJson, Set<String> expectedAttributeSet )
+    {
+        assertEquals( actualJson.getArray( "trackedEntityAttributes" ).size(), expectedAttributeSet.size() );
+        int i = 0;
+        while ( i < expectedAttributeSet.size() )
+        {
+            assertTrue( expectedAttributeSet.contains(
+                actualJson.getArray( "trackedEntityAttributes" ).get( i ).asObject().get( "displayName" ).asObject()
```

Comment:
@jbee maybe for this PR and also because I'd like to learn 😋 do you have a suggestion of how we could write this differently now that you added some new capabilities to json-tree?

My understanding is that we have a Set<String> of expected attribute `displayName`s and a JSON `trackedEntityAttributes` list. Each element in the list is a complex object with a field of `displayName`. We want to make sure that the `trackedEntityAttributes` contains all and only the displayNames in the `expectedAttributeSet`.

Reply:
Hm.. yes there are several things that are odd here.

First `get( ...).asObject()` should be written as `getObject(...)`. The `asObject()` is only meant to be used where we get a `JsonValue` handed without having control over the type but we expect it to be a `JsonObject` so we "cast" using `asObject()`. If we drill into a tree we can directly use `getObject(...).

Second, I'd assume the `displayName` is **not** an object but a string. So I'd expect to see `getString("displayName")`.

If a loop was needed here, it should be a `for` loop not a `while` - but I think we can just make an assert instead. 

If I read the test correct this could be a single `assertEquals`
```java
assertEquals( expectedAttributeSet, actualJson.getArray( "trackedEntityAttributes" )
  .viewAsList( e -> e.asObject().getString( "displayName" ) )
  .stream().map( JsonString::string )
  .collect( toSet() )
);
```
Key is the `viewAsList` where we to the equivalent of a `stream().map` just that this is purely a view transformation from the array of objects to the array of the object's display name. Then we have a `JsobList<JsonString>` of the display names which we convert to a `Set<String>` using `.stream().map( JsonString::string ).collect( toSet() )`.

This assert has the benefit of clearly telling you how the sets differ should that be the case. If that is not that important you can use `containsAll`. 

```java
assertTrue( actualJson.getArray( "trackedEntityAttributes" )
  .viewAsList( e -> e.asObject().getString( "displayName" ) )
  .containsAll(JsonString::string, expectedAttributeSet )
);
```
If hopefully becomes clear how this is doing the same just the other way around. Only the true/false is less helpful if it fails.